PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc

```diff
@@ -224,11 +240,35 @@ void pat::PATPackedCandidateProducer::produce(edm::Event& iEvent, const edm::Eve
           outPtrP->back().setAssociationQuality(pat::PackedCandidate::PVAssociationQuality(pat::PackedCandidate::UsedInFitTight));
         }
 	
-	if (puppiWeight.isValid()){
-	  reco::PFCandidateRef pkref( cands, ic );
-	  outPtrP->back().setPuppiWeight( (*puppiWeight)[pkref]);
-	  mappingPuppi[((*puppiCandsMap)[pkref]).key()]=ic;
-	}
+        if (puppiWeight.isValid()){
+           reco::PFCandidateRef pkref( cands, ic );
+                 // outPtrP->back().setPuppiWeight( (*puppiWeight)[pkref]);
+           
+           float puppiWeightVal = (*puppiWeight)[pkref];
+           float puppiWeightNoLepVal = 0.0;
+
+           // Check the "no lepton" puppi weights. 
+           // If present, then it is not a lepton, use stored weight
+           // If absent, it is a lepton, so set the weight to 1.0
+           if ( puppiWeightNoLep.isValid() ) {
+             // Look for the pointer inside the "no lepton" candidate collection.
+             auto pkrefPtr = pkref->sourceCandidatePtr(0);
+             bool foundNoLep = false;
+             for ( size_t ipcnl = 0; ipcnl < puppiCandsNoLepPtrs.size(); ipcnl++){
+              if (puppiCandsNoLepPtrs[ipcnl] == pkrefPtr){
+                foundNoLep = true;
+                  puppiWeightNoLepVal = puppiCandsNoLepPtrs[ipcnl]->pt()/cand.pt(); // a hack for now, should use the value map
```

Comment:
the previous code was picking pt from the puppiCandsNoLep, not from the source ptr.
If it matters functionally, you should probably have 
```c++
// define outside the loop: auto const& puppiWeightNoLepV = puppiWeightNoLep.product();
puppiWeightNoLepVal = puppiWeightNoLepV[ipcnl].pt()/cand.pt();
```